kbfsgit/runner.go

```diff
@@ -856,38 +848,60 @@ func (r *runner) handlePushBatch(ctx context.Context, args [][]string) (
 			return err
 		}
 
-		if !refspec.IsForceUpdate() {
-			r.log.CDebugf(
-				ctx, "Turning a non-force push into a force push for now: %s",
-				refspec)
-		}
-
-		start := strings.Index(push[0], ":") + 1
-		dst := push[0][start:]
-
 		// Delete the reference in the repo if needed; otherwise,
 		// fetch from the local repo into the remote repo.
 		if refspec.IsDelete() {
+			start := strings.Index(push[0], ":") + 1
+			dst := push[0][start:]
+
 			if refspec.IsWildcard() {
 				results[dst] = errors.Errorf(
 					"Wildcards not supported for deletes: %s", refspec)
 				continue
 			}
 			err = repo.Storer.RemoveReference(plumbing.ReferenceName(dst))
+			if err == gogit.NoErrAlreadyUpToDate {
+				err = nil
+			}
+			results[dst] = err
 		} else {
-			err = remote.FetchContext(ctx, &gogit.FetchOptions{
-				RemoteName: localRepoRemoteName,
-				RefSpecs:   []gogitcfg.RefSpec{refspec},
-				StatusChan: statusChan,
-			})
+			if !refspec.IsForceUpdate() {
+				r.log.CDebugf(ctx,
+					"Turning a non-force push into a force push for now: %s",
+					refspec)
+			}
+
+			refspecs = append(refspecs, refspec)
+		}
+		if err != nil {
+			r.log.CDebugf(ctx, "Error fetching %s: %+v", refspec, err)
+		}
+	}
+
+	if len(refspecs) > 0 {
+		var statusChan plumbing.StatusChan
+		if r.verbosity >= 1 {
+			s := make(chan plumbing.StatusUpdate)
+			defer close(s)
+			statusChan = plumbing.StatusChan(s)
+			go r.processGogitStatus(ctx, s)
 		}
+
+		err = remote.FetchContext(ctx, &gogit.FetchOptions{
+			RemoteName: localRepoRemoteName,
+			RefSpecs:   refspecs,
+			StatusChan: statusChan,
+		})
 		if err == gogit.NoErrAlreadyUpToDate {
 			err = nil
 		}
-		if err != nil {
-			r.log.CDebugf(ctx, "Error fetching %s: %+v", refspec, err)
+		// All refspecs in the batch get the same error.
```

Comment:
Only all pushes get the same error; deletes still have their own errors correct?

Reply:
True, I clarified the comment.